python/python/bystro/api/cli.py

```diff
@@ -0,0 +1,587 @@
+
+    Returns
+    -------
+    CachedAuth
+        The cached authentication state.
+    """
+    if print_result:
+        print(f"\nSigning up for Bystro with email: {args.email}, name: {args.name}")
+
+    fq_host = _fq_host(args)
+    url = f"{fq_host}/api/user"
+
+    data = {"email": args.email, "name": args.name, "password": args.password}
+
+    response = requests.put(url, data=data, timeout=30)
+
+    if response.status_code != 200:
+        raise RuntimeError(
+            f"Login failed with response status: {response.status_code}. Error: \n{response.text}\n"
+        )
+
+    res = mjson.decode(response.text, type=SignupResponse)
+    state = CachedAuth(
+        access_token=res.access_token,
+        url=fq_host,
+        email=args.email,
+    )
+
+    save_state(
+        state,
+        args.dir,
+        print_result,
+    )
+
+    if print_result:
+        print("\nSignup & authentication successful. You may now use the Bystro API!\n")
+
+    return state
+
+
+def login(args: argparse.Namespace, print_result=True) -> CachedAuth:
+    """
+    Logs in to the server and saves the authentication state to a file.
+
+    Parameters
+    ----------
+    args : argparse.Namespace
+        The arguments passed to the command.
+    print_result : bool, optional
+        Whether to print the result of the login operation, by default True.
+
+    Returns
+    -------
+    CachedAuth
+        The cached authentication state.
+    """
+    fq_host = _fq_host(args)
+
+    if print_result:
+        print(f"\nLogging into {fq_host} with email: {args.email}.")
+
+    url = f"{fq_host}/api/user/auth/local"
+
+    body = {"email": args.email, "password": args.password}
+
+    response = requests.post(url, data=body, timeout=30)
+
+    if response.status_code != 200:
+        raise RuntimeError(
+            f"Login failed with response status: {response.status_code}. Error: \n{response.text}\n"
+        )
+
+    res = mjson.decode(response.text, type=LoginResponse)
+    state = CachedAuth(access_token=res.access_token, url=fq_host, email=args.email)
+    save_state(state, args.dir, print_result)
+
+    if print_result:
+        print("\nLogin successful. You may now use the Bystro API!\n")
+
+    return state
+
+
+def authenticate(args) -> tuple[CachedAuth, dict]:
+    """
+    Authenticates the user and returns the url, auth header, and email.
+
+    Parameters
+    ----------
+    args : argparse.Namespace
+        The arguments passed to the command.
+
+    Returns
+    -------
+    tuple[CachedAuth, dict]
+        The cached auth credentials and auth header
+    """
+    state = load_state(args.dir)
+
+    if not state:
+        raise ValueError("\n\nYou are not logged in. Please login first.\n")
+
+    header = {"Authorization": f"Bearer {state.access_token}"}
+    return state, header
+
+def get_jobs(
+    args: argparse.Namespace, print_result=True
+) -> list[JobBasicResponse] | dict:
```

Comment:
I don't have time to model the job response when fetching job details (which is what happens when we fetch a single job), because that schema is flexible depending on what processing the job undergoes (ancestry, etc). Let's not block this PR on that, typing that in a way that doesn't restrict functionality will take a bit of time.